listenbrainz/labs_api/labs/api/recording_from_recording_mbid.py

```diff
@@ -54,45 +56,84 @@ def fetch(self, params, offset=-1, count=-1):
+                               ON rl.release_group = rg.id
+                             JOIN release_group_meta rgm
+                               ON rg.id = rgm.id
                             WHERE r.gid
                                IN %s
-                         GROUP BY r.gid, r.id, r.name, r.length, r.comment, ac.id, ac.name
+                         GROUP BY r.gid, r.id, r.name, r.length, r.comment, ac.id, ac.name, rgm.first_release_date_year
                          ORDER BY r.gid'''
 
                 args = [tuple([psycopg2.extensions.adapt(p) for p in mbids])]
-                if count > 0:
-                    query += " LIMIT %s"
-                    args.append(count)
-                if offset >= 0:
-                    query += " OFFSET %s"
-                    args.append(offset)
-
                 curs.execute(query, tuple(args))
 
-                output = []
+                # Build an index of all the fetched recordings and clean up UUID() -> str
+                recording_index = {}
                 while True:
                     row = curs.fetchone()
                     if not row:
                         break
 
                     r = dict(row)
-                    r['recording_mbid'] = str(r['recording_mbid'])
-                    r['[artist_credit_mbids]'] = [str(r) for r in r['artist_credit_mbids']]
+                    mbid = str(r['recording_mbid'])
+                    r['recording_mbid'] = mbid
+                    recording_index[mbid] = dict(row)
+
+                # Finally collate all the results, ensuring that we have one entry with original_recording_mbid for each 
+                # input argument
+                output = []
+                for p in params:
+                    mbid = p['[recording_mbid]'] 
+                    try:
+                        r = copy.copy(recording_index[mbid])
+                    except KeyError:
+                        try:
+                            r = copy.copy(recording_index[redirect_index[mbid]])
+                        except KeyError:
+                            out = {} 
+                            for k in self.outputs():
+                                if k == 'original_recording_mbid':
+                                    out['original_recording_mbid'] = mbid
+                                else:
+                                    out['original_recording_mbid'] = None
+                            output.append(out)
+                            continue
+
+                    r['[artist_credit_mbids]'] = list(set([str(r) for r in r['artist_credit_mbids']]))
                     del r['artist_credit_mbids']
+                    r['original_recording_mbid'] = mbid
                     output.append(r)
 
-        return output
+                if offset > 0 and count > 0:
+                    return output[offset:offset+count]
+
+                if offset > 0 and count < 0:
+                    return output[offset:]
```

Comment:
An issue with doing the limit in python and not in postgres is that you have to perform the whole query anyway to get the data. However, because you have to do the redirect lookup, I'm not sure if there's a way around this, right?
Any thoughts on a default count value?

Reply:
Sadly with the case that you brought up yesterday there is no longer a 1:1 relation between the number of rows requested by the caller and the number of rows requested by this query, making limit/count potentially harmful.

We could probably continue to tweak this function to make it more efficient -- however, I feel that we're in position where we are trying to make things work, rather than making them fast. While I am certain that this Element will be central to Troi in nearly all cases, I don't feel like right now is the time to hone this into perfection. I'll add a note in the comment to this affect and then move on.

Reply:
As for default count value, I'm actually keen on letting the user set count to the number of rows required -- as to avoid paginating right now. Another thing to think about as we move closer to production.